### PR TITLE
[EffectKaraoke] - Several bugs fixed + Update

### PR DESCRIPTION
**Pull request by ivandrofly**
Fixed 
    - (\<font color="#808080">MUSIC\</font>)
    - every \<\<constellation\>\>...
    - Do not include \r and \n when coloring e.g: \r\</font> or \n\</font>

Copy text above and run EffectKaraoke and see the result.

### CODE REVIEW

File: src/Forms/EffectKaraoke.cs [hunk 1]
```diff
@@ -258,16 +262,19 @@ private void MakeAnimation()
...
             while (i < _paragraph.Text.Length)
             {
                 var c = _paragraph.Text[i];
-                if (i == 0 && _paragraph.Text.StartsWith("{\\", StringComparison.Ordinal) && _paragraph.Text.IndexOf('}') > 2)
-                {
-                    int idx = _paragraph.Text.IndexOf('}');
-                    alignment = _paragraph.Text.Substring(0, idx + 1);
-                    i = idx;
-                }
-                else if (tagOn)
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]: Not necessary to be inside loop (Run once)

File: src/Forms/EffectKaraoke.cs [hunk 2]
```diff
@@ -225,10 +228,11 @@ private void PlayAnimation()
             timer1.Start();
         }
 
-        private static double CalculateStepLength(string text, double duration)
+        public static double CalculateStepLength(string text, double duration)
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]: Made public so Effect TypeWriter and call it (Patch for typewriter will be pull soon)

File: src/Forms/EffectKaraoke.cs [hunk 3]
```diff
@@ -85,11 +89,11 @@ private void AddToPreview(RichTextBox rtb, string text)
             int i = 0;
             while (i < text.Length)
             {
-                if (text[i] == '<')
+                if (text[i] == '<' && IsTagFollowIndex(text, i))
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]: Do not assume that all text that follows '<' is going to be html tag.

File: src/Forms/EffectKaraoke.cs [hunk 4]
```diff
@@ -36,7 +36,11 @@ private void FormEffectkaraoke_KeyDown(object sender, KeyEventArgs e)
 
         internal void Initialize(Paragraph paragraph)
         {
-            _paragraph = paragraph;
+            _paragraph = new Paragraph(paragraph);
+            if (_paragraph.Text.Length > 3)
+            {
+                _paragraph.Text = HtmlUtil.FixUpperTags(_paragraph.Text);
+            }
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]: Do all change in a copy instead (Allows us to revert if tags was indeed lowered)

File: src/Forms/EffectKaraoke.cs [hunk 5]
```diff
@@ -383,5 +389,28 @@ private void ButtonOkClick(object sender, EventArgs e)
             DialogResult = DialogResult.OK;
         }
 
+        public static bool IsTagFollowIndex(string text, int index)
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]: Will make sure that text following the '<' are indeed html tag

File: src/Forms/EffectKaraoke.cs [hunk 6]
```diff
@@ -306,12 +313,20 @@ private void MakeAnimation()
...
+                    // Take next char if current is (\r or \n);
+                    while ((_paragraph.Text[i] == '\r' || _paragraph.Text[i] == '\n') && (i + 1 < _paragraph.Text.Length))
+                    {
+                        text += _paragraph.Text[++i];
+                    }
                     string afterText = string.Empty;
                     if (i + 1 < _paragraph.Text.Length)
                         afterText = _paragraph.Text.Substring(i + 1);
-                    if (string.Compare(afterText, endFontTag, StringComparison.InvariantCultureIgnoreCase) == 0 && afterCurrentTag.StartsWith("<font", StringComparison.InvariantCultureIgnoreCase))
+                    if (afterText.StartsWith(endFontTag, StringComparison.Ordinal) && afterCurrentTag.StartsWith("<font", StringComparison.Ordinal))
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]: This is was what this line was causing
1
00:00:01,000 --> 00:00:01,285
<font color="#ff0000">(</font><font color="#808080">MUSIC</font>)

2
00:00:01,286 --> 00:00:01,570
<font color="#ff0000">(M</font><font color="#808080">USIC</font>)

3
00:00:01,571 --> 00:00:01,856
<font color="#ff0000">(MU</font><font color="#808080">SIC</font>)

4
00:00:01,857 --> 00:00:02,142
<font color="#ff0000">(MUS</font><font color="#808080">IC</font>)

5
00:00:02,143 --> 00:00:02,428
<font color="#ff0000">(MUSI</font><font color="#808080">C</font>)

6
00:00:02,429 --> 00:00:02,713
<font color="#ff0000">(MUSIC</font><font color="#808080"></font>)

7
00:00:02,714 --> 00:00:03,000
<font color="#ff0000">(MUSIC</font>)</font>


Note: Original text: (<font color="#808080">MUSIC</font>)

File: src/Forms/EffectKaraoke.cs [hunk 7]
```diff
@@ -330,17 +345,8 @@ private void MakeAnimation()
...
-
-            if (numericUpDownDelay.Value > 0)
-            {
-                startMilliseconds = index * stepsLength;
-                startMilliseconds += _paragraph.StartTime.TotalMilliseconds;
-                endMilliseconds = _paragraph.EndTime.TotalMilliseconds;
-                start = new TimeCode(startMilliseconds);
-                end = new TimeCode(endMilliseconds);
-                _animation.Add(new Paragraph(start, end, startFontTag + _paragraph.Text + endFontTag));
-            }
```
Comment (ivandrofly) [inline src/Forms/EffectKaraoke.cs]:
```
1101
01:04:43,380 --> 01:04:43,493
<font color="#ff0000">H</font>eh. And you knew everything.

1102
01:04:43,494 --> 01:04:43,607
<font color="#ff0000">He</font>h. And you knew everything.

1103
01:04:43,608 --> 01:04:43,722
<font color="#ff0000">Heh</font>. And you knew everything.

1104
01:04:43,723 --> 01:04:43,836
<font color="#ff0000">Heh.</font> And you knew everything.

1105
01:04:43,837 --> 01:04:43,950
<font color="#ff0000">Heh. </font>And you knew everything.

1106
01:04:43,951 --> 01:04:44,064
<font color="#ff0000">Heh. A</font>nd you knew everything.

1107
01:04:44,065 --> 01:04:44,178
<font color="#ff0000">Heh. An</font>d you knew everything.

1108
01:04:44,179 --> 01:04:44,292
<font color="#ff0000">Heh. And</font> you knew everything.

1109
01:04:44,293 --> 01:04:44,407
<font color="#ff0000">Heh. And </font>you knew everything.

1110
01:04:44,408 --> 01:04:44,521
<font color="#ff0000">Heh. And y</font>ou knew everything.

1111
01:04:44,522 --> 01:04:44,635
<font color="#ff0000">Heh. And yo</font>u knew everything.

1112
01:04:44,636 --> 01:04:44,749
<font color="#ff0000">Heh. And you</font> knew everything.

1113
01:04:44,750 --> 01:04:44,863
<font color="#ff0000">Heh. And you </font>knew everything.

1114
01:04:44,864 --> 01:04:44,977
<font color="#ff0000">Heh. And you k</font>new everything.

1115
01:04:44,978 --> 01:04:45,092
<font color="#ff0000">Heh. And you kn</font>ew everything.

1116
01:04:45,093 --> 01:04:45,206
<font color="#ff0000">Heh. And you kne</font>w everything.

1117
01:04:45,207 --> 01:04:45,320
<font color="#ff0000">Heh. And you knew</font> everything.

1118
01:04:45,321 --> 01:04:45,434
<font color="#ff0000">Heh. And you knew </font>everything.

1119
01:04:45,435 --> 01:04:45,548
<font color="#ff0000">Heh. And you knew e</font>verything.

1120
01:04:45,549 --> 01:04:45,662
<font color="#ff0000">Heh. And you knew ev</font>erything.

1121
01:04:45,663 --> 01:04:45,777
<font color="#ff0000">Heh. And you knew eve</font>rything.

1122
01:04:45,778 --> 01:04:45,891
<font color="#ff0000">Heh. And you knew ever</font>ything.

1123
01:04:45,892 --> 01:04:46,005
<font color="#ff0000">Heh. And you knew every</font>thing.

1124
01:04:46,006 --> 01:04:46,119
<font color="#ff0000">Heh. And you knew everyt</font>hing.

1125
01:04:46,120 --> 01:04:46,233
<font color="#ff0000">Heh. And you knew everyth</font>ing.

1126
01:04:46,234 --> 01:04:46,347
<font color="#ff0000">Heh. And you knew everythi</font>ng.

1127
01:04:46,348 --> 01:04:46,462
<font color="#ff0000">Heh. And you knew everythin</font>g.

1128
01:04:46,463 --> 01:04:46,576
<font color="#ff0000">Heh. And you knew everything</font>.

1129
01:04:46,577 --> 01:04:46,691
<font color="#ff0000">Heh. And you knew everything.</font>
```

Why reinserting `<font color="#ff0000">Heh. And you knew everything.</font>` twice?

**Comment by ivandrofly** [comment]
Please leave comment if you don't agree with something.

**Comment by niksedk** [comment]
Looks good :)
The "new line" bug regarding subtitles with two lines is fixed - and the tests I did worked fine too.

**Comment by ivandrofly** [comment]
:)

**Pull request merged**